### PR TITLE
Address some type-stability issues

### PR DESCRIPTION
**Pull request by willtebbutt**
We had a few type-stability issues. I've introduced tests and fixes that are largely inspired by Seth's work on [FiniteDifferences](https://github.com/JuliaDiff/FiniteDifferences.jl/pull/118)'s `to_vec` function.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -173,3 +173,4 @@ package doesn't currently support the functionality that you need.
 # Gotchas
 
 1. `Integer`s typically don't take part in the kind of optimisation procedures that this package is designed to handle. Consequently, `flatten(::Integer)` produces an empty vector.
+2. `deferred` has some type-stability issues when used in conjunction with abstract types. For example, `flatten(deferred(Normal, 5.0, 4.0))` won't infer properly. A simple work around is to write a function `normal(args...) = Normal(args...)` and work with `deferred(normal, 5.0, 4.0)` instead.
```
Comment (rofinn) [inline README.md:176]: Hmm, this is a bit awkward :(

**Comment by rofinn** [comment]
Hmm, maybe we should switch this repo over to GH actions?

**Comment by willtebbutt** [comment]
> Hmm, maybe we should switch this repo over to GH actions?

Yes, we should. Probably a job for tomorrow.

File: README.md (hunk 1 above)
Comment (willtebbutt) [reply README.md:176]: Yeah, it's a bit sad. I'm not really sure what to do about it for now, so I figure it's better to document the problem and the workaround until we find a nice solution.

Comment (rofinn) [reply README.md:176]: Out of curiosity, do you know if this holds true for even Julia nightly? I'm just wondering if we should open an issue against the base repo to try and handle this case?

**Review by rofinn — APPROVED** [review]
LGTM. It'd be nice to confirm that the `deferred` type instability issue is present for all julia releases. I'm guessing it probably is, but maybe we can open and issue and reference it here?

**Comment by willtebbutt** [comment]
@rofinn @oxinabox tests are failing on 1.3 because `cumsum` wasn't defined on `Tuple`s until 1.5.

I took a look at using `Compat` for it -- no luck unfortunately. Also looked at the definition in Base on 1.5 -- looks like a reasonable amount of the code to make this work just wasn't available pre-1.5. Any thoughts on the easiest way to proceed?

**Comment by rofinn** [comment]
FWIW, I'd probably just add a version check and do for `cumsum(collect(lengths))` for julia releases pre 1.5

**Pull request merged**